src/darwin/Framework/CHIP/MTROTAImageTransferHandler.mm

```diff
@@ -0,0 +1,402 @@
+
+                CHIP_ERROR err = mTransfer.AcceptTransfer(acceptData);
+                LogErrorOnFailure(err);
+                AsyncResponder::NotifyEventHandled(event, err);
+            }
+                          errorHandler:^(NSError *) {
+                              // Not much we can do here
+                          }];
+    };
+
+    auto nodeId = @(mNodeId.Value());
+
+    auto strongDelegate = mDelegate;
+    auto delegateQueue = mDelegateNotificationQueue;
+    if (strongDelegate == nil || delegateQueue == nil) {
+        LogErrorOnFailure(CHIP_ERROR_INCORRECT_STATE);
+        AsyncResponder::NotifyEventHandled(event, CHIP_ERROR_INCORRECT_STATE);
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    dispatch_async(delegateQueue, ^{
+        if ([strongDelegate respondsToSelector:@selector
+                            (handleBDXTransferSessionBeginForNodeID:controller:fileDesignator:offset:completion:)]) {
+            [strongDelegate handleBDXTransferSessionBeginForNodeID:nodeId
+                                                        controller:controller
+                                                    fileDesignator:fileDesignator
+                                                            offset:offset
+                                                        completion:completionHandler];
+        } else {
+            [strongDelegate handleBDXTransferSessionBeginForNodeID:nodeId
+                                                        controller:controller
+                                                    fileDesignator:fileDesignator
+                                                            offset:offset
+                                                 completionHandler:completionHandler];
+        }
+    });
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR MTROTAImageTransferHandler::OnTransferSessionEnd(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(mFabricIndex.HasValue(), CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mNodeId.HasValue(), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    if (event.EventType == TransferSession::OutputEventType::kTransferTimeout) {
+        error = CHIP_ERROR_TIMEOUT;
+    } else if (event.EventType != TransferSession::OutputEventType::kAckEOFReceived) {
+        error = CHIP_ERROR_INTERNAL;
+    }
+
+    auto * controller = [[MTRDeviceControllerFactory sharedInstance] runningControllerForFabricIndex:mFabricIndex.Value()];
+    VerifyOrReturnError(controller != nil, CHIP_ERROR_INCORRECT_STATE);
+    auto nodeId = @(mNodeId.Value());
+
+    auto strongDelegate = mDelegate;
+    auto delegateQueue = mDelegateNotificationQueue;
+    if (strongDelegate == nil || delegateQueue == nil) {
+        error = CHIP_ERROR_INCORRECT_STATE;
+        LogErrorOnFailure(error);
+        AsyncResponder::NotifyEventHandled(event, error);
```

Comment:
I will modify the OnTransferSessionEnd/OnTransferSessionBegin API to return an error and then the caller of this will call the NotifyEventHandled. That should avoid the double calls.

Reply:
Contract between onTransferSession*** and HandleTransferSessionOutput should be :
1. If callee returns error, then HandleTransferSessionOutput should call NotifyEventHandled. Callee does not call it. (returns error)
2. If callee returns success, it must call NotifyEventHandled (possibly async)

Reply:
done.